blockchain/v2/reactor_test.go

```diff
@@ -1,22 +1,552 @@
+			},
+		},
+		{
+			name:   "termination on max peer height - two peers, noBlock error",
+			params: params,
+			msgs: []testEvent{
+				{"AddPeer", "P1", nil},
+				{"AddPeer", "P2", nil},
+				{"Receive", "P1", bcStatusResponseMessage{Height: 13}},
+				{"Receive", "P2", bcStatusResponseMessage{Height: 15}},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P1", bcNoBlockResponseMessage{11}},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(12)}},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(11)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(13)}},
+				{"Process", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(14)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(15)}},
+				{"Process", "", nil},
+			},
+		},
+		{
+			name:   "termination on max peer height - two peers, remove one peer",
+			params: params,
+			msgs: []testEvent{
+				{"AddPeer", "P1", nil},
+				{"AddPeer", "P2", nil},
+				{"Receive", "P1", bcStatusResponseMessage{Height: 13}},
+				{"Receive", "P2", bcStatusResponseMessage{Height: 15}},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"RemovePeer", "P1", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(12)}},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(11)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(13)}},
+				{"Process", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(14)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(15)}},
+				{"Process", "", nil},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			reactor := newTestReactor(params)
+			reactor.Start()
+			mockSwitch := &mockSwitchIo{switchedToConsensus: false}
+			reactor.io = mockSwitch
+			// time for go routines to start
+			time.Sleep(time.Millisecond)
+
+			for i := 0; i < len(tt.msgs); i++ {
+				step := tt.msgs[i]
+				switch step.evType {
+				case "AddPeer":
+					reactor.AddPeer(mockPeer{id: p2p.ID(step.peer)})
+				case "RemovePeer":
+					reactor.RemovePeer(mockPeer{id: p2p.ID(step.peer)}, fmt.Errorf("some error"))
+				case "Receive":
+					reactor.Receive(channelID, mockPeer{id: p2p.ID(step.peer)}, cdc.MustMarshalBinaryBare(step.event))
+				case "BlockReq":
+					reactor.scheduler.send(trySchedule{time: time.Now()})
+				case "Process":
+					reactor.processor.send(pcProcessBlock{})
+				}
+				// time for messages to propagate between routines
+				time.Sleep(time.Millisecond)
+			}
+
+			time.Sleep(20 * time.Millisecond)
```

Comment:
Seems like we should just have a reactor.Wait() method which will return when the demuxer hits the final event.